classes/table/accessaudit_table.php

```diff
@@ -0,0 +1,243 @@
+
+
+defined('MOODLE_INTERNAL') || die;
+
+require_once($CFG->libdir . '/tablelib.php');
+
+class accessaudit_table extends \table_sql implements \renderable {
+
+    /**
+     * A list of filters to be applied to the sql query.
+     * @var \stdClass
+     */
+    protected $filters;
+
+    /**
+     * A current page number.
+     * @var int
+     */
+    protected $page;
+
+    public function __construct($uniqueid, \moodle_url $url, $filters = array(), $download = '', $page = 0, $perpage = 100) {
+
+        parent::__construct($uniqueid);
+
+        $this->pagesize = $perpage;
+        $this->page = $page;
+        $this->filters = (object)$filters;
+
+        // Define columns in the table.
+        $this->define_table_columns();
+
+        // Define configs.
+        $this->define_table_configs($url);
+
+        // Set download status.
+        $this->is_downloading($download, 'accessaudit_accessaudit_report');
+
+    }
+
+    /**
+     * Setup the headers for the html table.
+     */
+    protected function define_table_columns() {
+        $cols = array(
+            'idnumber' => get_string('idnumber'),
+            'username' => get_string('username'),
+            'email' => get_string('email'),
+            'fullname' => get_string('fullname'),
+            'role' => get_string('role'),
+            'context' => get_string('report_context', 'report_accessaudit'),
+            'contextpathraw' => get_string('report_contextpathraw', 'report_accessaudit'),
+            'contextpathhuman' => get_string('report_contextpathhuman', 'report_accessaudit')
+        );
+
+        $this->define_columns(array_keys($cols));
+        $this->define_headers(array_values($cols));
+    }
+
+    /**
+     * Define table configs.
+     *
+     * @param \moodle_url $url url of the page where this table would be displayed.
+     */
+    protected function define_table_configs(\moodle_url $url) {
+        $urlparams = (array)$this->filters;
+
+        unset($urlparams['submitbutton']);
+
+        $url->params($urlparams);
+        $this->define_baseurl($url);
+
+        // Set table configs.
+        $this->collapsible(false);
+        $this->sortable(false);
+        $this->pageable(false);
+
+        $this->is_downloadable(true);
+        $this->show_download_buttons_at([TABLE_P_BOTTOM]);
+    }
+
+    /**
+     * Query the reader. Store results in the object for use by build_table.
+     *
+     * @param int $pagesize size of page for paginated displayed table.
+     * @param bool $useinitialsbar do you want to use the initials bar.
+     */
+    public function query_db($pagesize, $useinitialsbar = true) {
+        global $DB;
+
+        $offset = $pagesize * $this->page;
+        $limit = $pagesize;
+
+        list($countsql, $countparams) = $this->get_sql_and_params(true);
+        list($sql, $params) = $this->get_sql_and_params();
+
+        $total = $DB->count_records_sql($countsql, $countparams);
+
+        if ($this->is_downloading()) {
```

Comment:
So this code down to line 147 is roughly:

1) open a recordset for streaming data from db
2) iterate through that recordset and copy it all into memory
3) including calling \context_helper::instance_by_id on every row

When you are dealing with massive data this streaming won't work, it will consume more and more memory the large the data gets.

The way to do this with fixed memory overhead is just return the data stream directly like:

https://github.com/moodle/moodle/blob/master/mod/feedback/classes/responses_table.php#L363-L367

and then augment the data 'just in time' like this for the columns that need it:

https://github.com/moodle/moodle/blob/master/mod/feedback/classes/responses_table.php#L201-L210

Reply:
Hi @brendanheywood 
This is improved in commit c9d1841e22374e2e6d31df555e05f79bc195e872
Please let me know if that is correct
Thanks for your review

Reply:
looks great